in_toto/keylib.go

```diff
@@ -2,220 +2,402 @@ package in_toto
+	switch key.KeyType {
+	case "rsa", "ecdsa":
+		// pam.Decode returns the parsed pem block and a rest.
+		// The rest is everything, that could not be parsed as PEM block.
+		// Therefore we can drop this via using the blank identifier "_"
+		data, _ := pem.Decode([]byte(key.KeyVal.Private))
+		if data == nil {
+			return signature, ErrNoPEMBlock
+		}
+		parsedKey, err := ParseKey(data.Bytes)
+		if err != nil {
+			return signature, err
+		}
+		hashed := sha256.Sum256(signable)
+		switch parsedKey.(type) {
+		case *rsa.PrivateKey:
+			if key.KeyType != "rsa" {
+				return signature, ErrInvalidKeyType
+			}
+			// We use rand.Reader as secure random source for rsa.SignPSS()
+			signatureBuffer, err = rsa.SignPSS(rand.Reader, parsedKey.(*rsa.PrivateKey), crypto.SHA256, hashed[:],
+				&rsa.PSSOptions{SaltLength: sha256.Size, Hash: crypto.SHA256})
+			if err != nil {
+				return signature, err
+			}
+		case *ecdsa.PrivateKey:
+			if key.KeyType != "ecdsa" {
+				return signature, ErrInvalidKeyType
+			}
+			// ecdsa.Sign returns a signature that consists of two components called: r and s
+			// We assume here, that r and s are of the same size nLen and that
+			// the signature is 2*nLen. Furthermore we must note  that hashes get truncated
+			// if they are too long for the curve. We use SHA256 for hashing, thus we should be
+			// ok with using the FIPS186-3 curves P256, P384 and P521.
+			r, s, err := ecdsa.Sign(rand.Reader, parsedKey.(*ecdsa.PrivateKey), hashed[:])
+			if err != nil {
+				return signature, nil
+			}
+			// Generate the ecdsa signature on the same way, as we do in the securesystemslib
+			// We construct a
+			signatureBuffer, err = asn1.Marshal(EcdsaSignature{
+				R: r,
+				S: s,
+			})
+		default:
+			return signature, fmt.Errorf("%w: %T", ErrUnsupportedKeyType, parsedKey)
+		}
+	case "ed25519":
+		privateHex, err := hex.DecodeString(key.KeyVal.Private)
+		if err != nil {
+			return signature, err
+		}
+		// Note: We can directly use the key for signing and do not
+		// need to use ed25519.NewKeyFromSeed().
+		signatureBuffer = ed25519.Sign(privateHex, signable)
+	default:
+		return signature, fmt.Errorf("%w: %s", ErrUnsupportedKeyType, key.KeyType)
 	}
-	privkey := ed25519.NewKeyFromSeed(seed)
-	signatureBuffer := ed25519.Sign(privkey, signable)
-
 	signature.Sig = hex.EncodeToString(signatureBuffer)
 	signature.KeyId = key.KeyId
-
 	return signature, nil
 }
+
+/*
+VerifySignature will verify unverified byte data via a passed key and signature.
+Supported key types are:
+
+	* RSA
+	* ED25519
+
+When encountering a RSA key, VerifySignature will decode the PEM block in the key
+and will call rsa.VerifyPSS() for verifying the RSA signature.
+When encountering an ed25519 key, Verifysignature will decode the hex string encoded
+public key and will use ed25519.Verify() for verifying the ed25519 signature.
+On success it will return nil. In case of an unsupported key type or any other error
+it will return an error.
```

Comment:
We should also take a look at the signature scheme of the key and fail if it is not one we support for verifying signatures.

Reply:
This should be done in the latest commit.